### PR TITLE
Speed up the function `min_max_string`

### LINKED ISSUE #1373: Improve performance of `min` and `max` aggregation kernels without nulls
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
The function `min_max_string` does some unnecessary checks when `null_count > 0 `. For example, we don't need to check `has_value` in every loop because it will always be `true` after the first loop. 
```rust
    if null_count == 0 {
        n = array.value(0);
        for i in 1..data.len() {
            let item = array.value(i);
            if cmp(n, item) {
                n = item;
            }
        }
    } else {
        n = "";
        let mut has_value = false;

        for i in 0..data.len() {
            let item = array.value(i);
            if data.is_valid(i) && (!has_value || cmp(n, item)) {
                has_value = true;
                n = item;
            }
        }
    }
```
https://github.com/apache/arrow-rs/blob/master/arrow/src/compute/kernels/aggregate.rs#L55-L64

Apart from that, I want this function to be cleaned up because the "for loops" here are not pretty.

**Describe the solution you'd like**
1. Performance should be improved when `null_count > 0`
2. No performance penalty is introduced in other cases
3. clean up the code. Maybe use some FP skills

**Describe alternatives you've considered**
We can also replace `array.value(i)` by `array.value_unchecked(i)`. But it will introduce some "unsafe", so I am not sure.

### PR DESCRIPTION
**Pull request by HaoYang670**
# Which issue does this PR close?
Closes #1373 .

# Rationale for this change
 1. Improve performance 
 2. clean up the code

# What changes are included in this PR?
1. Performance is improved when `null_count > 0`. See benchmark result.  
2. more functional styles and fewer mutable variables

## benchmark
```shell
min string 512          time:   [1.3687 us 1.3691 us 1.3696 us]                            
                        change: [-7.7279% -7.5833% -7.4242%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  1 (1.00%) high mild
  8 (8.00%) high severe

min nulls string 512    time:   [1.6630 us 1.6647 us 1.6665 us]                                  
                        change: [-33.478% -33.362% -33.244%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe
```

# Are there any user-facing changes?
No

### CODE REVIEW

**Comment by HaoYang670** [comment]
# Benchmark
## Before
```bash
min string 512          time:   [1.4824 us 1.4914 us 1.5094 us]                            
                        change: [-2.4834% -0.7235% +0.5457%] (p = 0.44 > 0.05)
                        No change in performance detected.
Found 13 outliers among 100 measurements (13.00%)
  2 (2.00%) high mild
  11 (11.00%) high severe

min nulls string 512    time:   [2.4456 us 2.4528 us 2.4610 us]                                  
                        change: [-0.7299% -0.2597% +0.2043%] (p = 0.26 > 0.05)
                        No change in performance detected.
Found 9 outliers among 100 measurements (9.00%)
  7 (7.00%) high mild
  2 (2.00%) high severe
```

## After
```bash
min string 512          time:   [1.5539 us 1.5552 us 1.5565 us]                            
                        change: [-0.4238% -0.2174% -0.0106%] (p = 0.04 < 0.05)
                        Change within noise threshold.
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) high mild
  4 (4.00%) high severe

min nulls string 512    time:   [1.8519 us 1.8536 us 1.8552 us]                                  
                        change: [-1.0138% -0.7613% -0.4859%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
  1 (1.00%) high severe
```

**Comment by HaoYang670** [comment]
I have tested that if we replace `array.value` by `array.value_unchecked`, there will be 10% extra performance improvement on the benchmark `min string 512`.  But it also introduces an unsafe block.

File: arrow/src/compute/kernels/aggregate.rs [hunk 1]
```diff
@@ -32,39 +32,26 @@ fn is_nan<T: ArrowNativeType + PartialOrd + Copy>(a: T) -> bool {
...
-        let mut has_value = false;
-
-        for i in 0..data.len() {
-            let item = array.value(i);
-            if data.is_valid(i) && (!has_value || cmp(n, item)) {
-                has_value = true;
-                n = item;
-            }
-        }
+        array
```
Comment (Dandandan) [inline arrow/src/compute/kernels/aggregate.rs:50]: According to your benchmark result, this new way a bit slower than before for nulls?

Comment (HaoYang670) [reply arrow/src/compute/kernels/aggregate.rs:50]: No. You can see the benchmark "min nulls string 512". Time decreases from 2.4 to 1.8. I this that is a performance improvement.

Comment (HaoYang670) [reply arrow/src/compute/kernels/aggregate.rs:50]: Also for the benchmark "min string 512 ", there is no big performance penalty introduced.

Comment (HaoYang670) [reply arrow/src/compute/kernels/aggregate.rs:50]: @Dandandan

Comment (Dandandan) [reply arrow/src/compute/kernels/aggregate.rs:50]: I think we could speed it up a bit more if we could iterate fast on the non-null values.

**Review by Dandandan — APPROVED** [review]

File: arrow/src/compute/kernels/aggregate.rs (hunk 1 above)
Comment (HaoYang670) [reply arrow/src/compute/kernels/aggregate.rs:50]: Ok, let me try the `array.value_unchecked()`

**Comment by HaoYang670** [comment]
```console
min string 512          time:   [1.3687 us 1.3691 us 1.3696 us]                            
                        change: [-7.7279% -7.5833% -7.4242%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  1 (1.00%) high mild
  8 (8.00%) high severe

min nulls string 512    time:   [1.6630 us 1.6647 us 1.6665 us]                                  
                        change: [-33.478% -33.362% -33.244%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe
```
Using `value_unchecked` can get 8% extra speedup for not-null values

**Comment by alamb** [comment]
Thanks @HaoYang670  and @Dandandan

**Pull request merged**